### PR TITLE
Migrate to typescript

### LINKED ISSUE #42: Migrate to Typescript

### PR DESCRIPTION
**Pull request by louisinger**
This PR migrates the whole project to typescript using https://github.com/bitjson/typescript-starter

TL;DR
* scripts to build and test are unchanged.
* removed the doc in README, use typedoc instead `yarn doc`.
* export the `Secp256k1ZKP` interfaces in order to let other libs import them if needed.
* split the functions into several files `generator.ts`, `pedersen.ts`, `ecc.ts` etc...
* No breaking changes in the API neither in the usage of the lib (still export an async function returning the lib)

it closes #42

### CODE REVIEW

**Comment by tiero** [comment]
We would love also a quick feedback from you as well @michael1011

**Comment by tiero** [comment]
I would mention if the API got any breaking changes or not @louisinger

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  "parserOptions": { "project": "./tsconfig.json" },
+  "env": { "es6": true },
+  "ignorePatterns": ["node_modules", "build", "coverage"],
+  "plugins": ["import", "eslint-comments"],
+  "extends": [
+    "eslint:recommended",
+    "plugin:eslint-comments/recommended",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:import/typescript",
+        "prettier",
```
Comment (michael1011) [inline .eslintrc.json:13]:
```suggestion
    "prettier",
```

File: .gitignore [hunk 2]
```diff
@@ -1,8 +1,9 @@
...
+src/**.js
 coverage
-build
-./test
-.vscode
-settings.json
-**/*bundle*
+*.log
+package-lock.json
+dist
```
Comment (michael1011) [inline .gitignore:9]: Some files are missing an EOF new line. Not a big deal but it's kinda bugging me :smile:

**Review by michael1011 — COMMENTED** [review]
Did not test it, but the approach looks good to me. Much nicer structure of the codebase and no more annoying writing of the types by hand.

**Pull request merged**